File: platforms/ios/Lists/Core/Storage/FrontmatterCodec.swift
import Foundation
import Yams

/// Encode/decode an `Item` to/from a markdown file with YAML frontmatter.
///
/// On-disk layout per file:
///
///     ---
///     id: 01HX...
///     type: task
///     title: Pay phone bill
///     ...
///     ---
///     <markdown body>
///
public enum FrontmatterCodec {
    public enum Error: Swift.Error, Equatable {
        case missingOpener
        case missingCloser
    }

    public static func encode(_ item: Item) throws -> String {
        let encoder = YAMLEncoder()
        encoder.options.sortKeys = false
        let yaml = try encoder.encode(item)
        let trimmed = yaml.hasSuffix("\n") ? yaml : yaml + "\n"
        let body = item.body.hasSuffix("\n") ? item.body : item.body + (item.body.isEmpty ? "" : "\n")
        return "---\n" + trimmed + "---\n" + body
    }

    public static func decode(_ source: String) throws -> Item {
        let parts = try splitFrontmatter(source)
        var item = try YAMLDecoder().decode(Item.self, from: parts.frontmatter)
        item.body = parts.body
        return item
    }

    private static func splitFrontmatter(_ source: String) throws -> (frontmatter: String, body: String) {
        // Accept files starting with "---\n" or BOM + "---\n"
        let stripped = source.trimmingPrefix("\u{FEFF}")
        guard stripped.hasPrefix("---\n") else { throw Error.missingOpener }
        let after = String(stripped.dropFirst(4))
        guard let closeRange = after.range(of: "\n---\n") else {
            // Allow file ending right at "\n---" (no trailing newline + body)
            if let endRange = after.range(of: "\n---", options: .backwards),
               endRange.upperBound == after.endIndex {
                return (String(after[..<endRange.lowerBound]), "")
            }
            throw Error.missingCloser
        }
        let frontmatter = String(after[..<closeRange.lowerBound])
        let body = String(after[closeRange.upperBound...])
        return (frontmatter, body)
    }
}

private extension String {
    func trimmingPrefix(_ prefix: String) -> String {
        hasPrefix(prefix) ? String(dropFirst(prefix.count)) : self
    }
}

File: platforms/android/core/src/main/kotlin/io/github/saxonbobart/lists/core/storage/FrontmatterCodec.kt
package io.github.saxonbobart.lists.core.storage

import io.github.saxonbobart.lists.core.model.Item

/**
 * Encode/decode an [Item] to/from a markdown file with YAML frontmatter —
 * the same on-disk layout the iOS app reads and writes:
 *
 * ```
 * ---
 * id: 01HX...
 * type: task
 * title: Pay phone bill
 * ...
 * ---
 * <markdown body>
 * ```
 */
object FrontmatterCodec {

    fun encode(item: Item): String {
        val yaml = ItemYaml.encode(item) // always newline-terminated
        val body = when {
            item.body.isEmpty() -> ""
            item.body.endsWith("\n") -> item.body
            else -> item.body + "\n"
        }
        return "---\n$yaml---\n$body"
    }

    fun decode(source: String): Item {
        val (frontmatter, body) = splitFrontmatter(source)
        return ItemYaml.decode(frontmatter).copy(body = body)
    }

    private fun splitFrontmatter(source: String): Pair<String, String> {
        // Accept files starting with "---\n" or BOM + "---\n"
        val stripped = source.removePrefix("\uFEFF")
        if (!stripped.startsWith("---\n")) throw YamlCodecException("Missing frontmatter opener")
        val after = stripped.substring(4)
        val close = after.indexOf("\n---\n")
        if (close < 0) {
            // Allow file ending right at "\n---" (no trailing newline + body)
            if (after.endsWith("\n---")) {
                return after.substring(0, after.length - 4) to ""
            }
            throw YamlCodecException("Missing frontmatter closer")
        }
        return after.substring(0, close) to after.substring(close + 5)
    }
}
